tim_common/markupmodels.py

```diff
@@ -77,11 +77,15 @@ class AccessField:
 @dataclass
 class ModelAnswerInfo:
     answer: str | None | Missing = missing
+    dueDate: PluginDateTime | datetime | None | Missing = (
+        missing  # TODO: this gets nullified when passed as str to external plugins
```

Comment:
Joitain plugineita kutsuessa (esim pali ja textfield) tämä taitaa muuttua stringiksi, ja sitten pluginia kutsuessa se palaa nullina. Saako tuota jotenkin kierrettyä? Jos tuohon laittaa | str, niin silloin se ei valita kunnolla virheistä vaan hyväksyy epävalidit ajat. Jos ei, niin laitan tuon /taskInfo-reitin kautta (sitä kautta taitaa tulla tuon nykyisen deadline-attribuutin tiedot answerbrowserille)

![kuva](https://user-images.githubusercontent.com/63715895/180016231-711822ac-5574-44f1-8352-6ff39868e8fa.png)

Esim cspluginissa (eka tehtävä) tuo toimii normaalisti, textfieldissä (vika tehtävä) ei: https://timdevs01-3.it.jyu.fi/view/modelanswer-duedate

Reply:
Hmm, tarkemmin katsottuna ilmeisesti `tim_common`issa on määritelty oma JSON-koodaaja:

https://github.com/TIM-JYU/TIM/blob/d6aec86596a2edfeba1d7d0b1f88c06ff7b515d8/tim_common/pluginserver_flask.py#L43-L48

ja se on oletusarvo plugineille

https://github.com/TIM-JYU/TIM/blob/d6aec86596a2edfeba1d7d0b1f88c06ff7b515d8/tim_common/pluginserver_flask.py#L224-L225

Jotkut pluginit näyttävät käyttävän [`TimJsonEncoder`](https://github.com/TIM-JYU/TIM/blob/aa111034968a6018b4824accaafdddc1f815b7a1/timApp/document/timjsonencoder.py#L12)ia, joka osaa käsitellä päivämäärät oikein. Eli todennäköisemmin tapahtuu nyt niin, että `PluginJsonEncoder` ei vaan käsittele päivämäärät.

Helpoin korjaus toistaiseksi olisi varmaan sellainen, että kopioi osan päivämäärän käsittelykoodista `tim_common`iin. Ainoastaan `if isinstance(o.__class__, DeclarativeMeta)`-käsittelyn pitäisi jättää kopioimatta, koska se vaatii SQLAlchemy-korjastoa, jota ei ole kaikissa konteissa (esim. cspluginissa sitä ei ole).
Parempi ratkaisu olisi muuttaa tuota `isinstance`-tarkistusta siten, että siinä ei olisi "kovaa" riippuvutta SQLAlchemyyn. Silloin `TimJsonEncoder` voisi siirtää `tim_common`iin ja käyttää kaikissa plugineissa.
Se pitäisi olla ihan mahdollinen, mutta en osaa näin äkkiä sanoa oikea tapa tehdä se Pythonissa. Varmaan kyllä sen saisi niin, että `o.__class__` muuttaa joksikin merkkijonoksi, josta voi tarkistaa, vastaako se `DeclarativeMeta`-luokkaa. Muutenhan tuon if:n sisällä oleva tavara ei näytä riippuvan SQLAlchemysta.